File: d3/D3_mvBase_TechDays/D3_PHP_Rest/Invoice.php
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN"
"http://www.w3.org/TR/html4/strict.dtd">

<?php
		// ****************************** retrieve from previous page a action's link arguments
        if(isset($_GET["invoiceNo"])) {
                $invoiceNo = $_GET["invoiceNo"];
        }
        elseif(isset($_POST["invoiceNo"])) {
                $invoiceNo = $_POST["invoiceNo"];
        }
        else {
                $invoiceNo = "";
        }
		//***** build up URI for calling Web Service to get specific invoice *****
        $uri  = "http://den-l-dg01:8181"; // replace with your host or leverage a config file or similar approach
        $uri .= "/Invoices";
        $uri .= "/getInvoice";
        $uri .= "?INVOICE_NO=".$invoiceNo;
        $jsonStr = file_get_contents($uri);
        $jsonArr = json_decode($jsonStr, true);
        $trxMst = $jsonArr["getInvoice"]["INVOICE_ITEM"]; // offset to get to the data structure in the Jason results *****
?>
 <html>
        <link href="acme.css" rel="stylesheet" type="text/css" />
        <head><title>Acme Invoice</title></head>
        <body   marginwidth="0"
                marginheight="0">
        <div id="InvoiceTopDiv">
        <img align="left"
                src="/img/acme.png"
                border="0">
				<FORM action="index.htm" method="post">
				<br clear="all">	
					<div id="InvoiceActions"><INPUT type="submit" value="New Search"> </div>
				</FORM>
        </div>
 
<table width=810px border=1>
        <tr><td valign="top"><table width=450px border=0>
<?php
		// ************************* create table and populate with the single valued attributes from the Jason Structure******
        echo "<tr><td width=80px>Bill to:</td>";
                echo "<td>";
                echo $trxMst['billtoCompanyName'];
                echo "</td></tr>";
        echo "<tr><td></td>";
                echo "<td>";
                echo $trxMst['billtoAddress'];
                echo "</td></tr>";
        echo "<tr><td></td>";
                echo "<td>";
                echo $trxMst['billtoCity'].", ";
                echo $trxMst['billtoState']." ";
                echo $trxMst['billtoZip'];
                echo "</td></tr>";
        echo "<tr><td>Ship to:</td>";
                echo "<td>";
                echo $trxMst['shiptoName'];
                echo "</td></tr>";
        echo "<tr><td></td>";
                echo "<td>";
                echo $trxMst['shiptoAddress'];
                echo "</td></tr>";
        echo "<tr><td></td>";
                echo "<td>";
                echo $trxMst['shiptoCity'].", ";
                echo $trxMst['shiptoState']." ";
                echo $trxMst['shiptoZip'];
                echo "</td></tr>";
        echo "<tr><td>Contact:</td>";
                echo "<td>";
                echo $trxMst['billtoContactName'];
                echo "</td></tr>";
        echo "<tr><td>Sales Rep:</td>";
                echo "<td>";
                echo $trxMst['salesRepName'];
                echo "</td></tr>";
        echo "<tr><td>Rep Phone:</td>";
                echo "<td>";
                echo $trxMst['salesRepPhone'];
                echo "</td></tr>";
?>
        </table></td>
        <td valign="top"><table width=350px border=0>
<?php
		// ************************* create second table and populate with the single valued attributes from the Jason Structure******
        echo "<tr><td width=80px>Invoice#:</td>";
                echo "<td width=240px>";
                echo $trxMst['invoiceNo'];
                echo "</td></tr>";
        echo "<tr><td>Account#:</td>";
                echo " <td>";
                echo $trxMst['custAccount'];
                echo "</td></tr>";
        echo "<tr><td>Ordered:</td>";
                echo " <td>";
                echo $trxMst['orderDate'];
                echo "</td></tr>";
        echo "<tr><td>Invoiced:</td>";
                echo " <td>";
                echo $trxMst['invoiceDate'];
                echo "</td></tr>";
        echo "<tr><td>Terms:</td>";
                echo " <td>";
                echo $trxMst['terms'];
                echo "</td></tr>";
        echo "<tr><td>Note:</td>";
                echo " <td>";
                echo $trxMst['comment'];
                echo "</td></tr>";
        echo "<tr><td>Gross:</td>";
                echo " <td>";
                echo $trxMst['invoiceGross'];
                echo "</td></tr>";
        echo "<tr><td>Discount:</td>";
                echo " <td>";
                echo $trxMst['invoiceDiscount'];
                echo "</td></tr>";
        echo "<tr><td>Net:</td>";
                echo " <td>";
                echo $trxMst['invoiceNet'];
                echo "</td></tr>";
?>
        </table></td></tr>
</table>
<br>

<!-- Create HTML table headers for invoice item list table -->

<table  width="750" border=1>
        <tr><th width=50px>Ln</th>
                <th width=50px>Qty</th>
                <th width=50px>Item#</th>
                <th width=350px>Description</th>
                <th width=75px>Unit Price</th>
                <th width=75px>Ext Price</th>
                <th width=50px>Disc</th>
                <th width=50px>Ext</th>
        </tr>
<?php
		// *****************  Loop through json array for each item in the invoice ***************
        $lineNo = 0;
        foreach ($trxMst["line"] as $line) {
                echo "<tr><td align='center'>".++$lineNo."</td>";
                echo "<td align='center'>".$line["lineQty"]."</td>";
                echo "<td align='center'>".$line["lineProductNo"]."</td>";
                echo "<td align='left'>".$line["lineDescription"]."</td>";
                echo "<td align='right'>".$line["lineUnitPrice"]."</td>";
                echo "<td align='right'>".$line["lineGross"]."</td>";
                echo "<td align='right'>".$line["lineDiscount" ]."</td>";
                echo "<td align='right'>".$line["lineExtNet"]."</td>";
                echo "</tr>";
        }
?>
</table>
        </body>
</html>

File: d3/D3_mvBase_TechDays/Visual Studio/WindowsFormsApplication_Acme_Rest/WindowsFormsApplication_Acme_Rest/Invoice.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.IO;
using System.Net;
using System.Web;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace WindowsFormsApplication_Acme_Rest
{
    public partial class Invoice : Form
    {
        public Invoice(String InvoiceSelection)
        {
            InitializeComponent();

            StartPosition = FormStartPosition.CenterScreen; //center form

            this.FormClosing += Invoice_FormClosing; // set close event

            // Create the web request  
            string uri = "http://den-vm-dg02:8181/Invoices/getInvoice?INVOICE_NO="; // substitute with your host path. You may want to read this in from a config file (not hardcoded)
            uri = uri + InvoiceSelection;

            using (WebClient webClient = new System.Net.WebClient())
            {
                WebClient n = new WebClient();
                var json = n.DownloadString(uri);
                JObject jObj = (JObject)JsonConvert.DeserializeObject(json);

                AccountNo.Text =jObj.SelectToken("getInvoice.INVOICE_ITEM.custAccount").ToString();
                BillTo.Text =   jObj.SelectToken("getInvoice.INVOICE_ITEM.billtoCompanyName").ToString();
                BillAddr.Text = jObj.SelectToken("getInvoice.INVOICE_ITEM.billtoAddress").ToString();

                var BillCity =  jObj.SelectToken("getInvoice.INVOICE_ITEM.billtoCity").ToString();
                var BillState = jObj.SelectToken("getInvoice.INVOICE_ITEM.billtoState").ToString();
                var BillZip =   jObj.SelectToken("getInvoice.INVOICE_ITEM.billtoZip").ToString();
                Bill_CityStateZip.Text = BillCity + ", " + BillState + " " + BillZip;

                Contact.Text =  jObj.SelectToken("getInvoice.INVOICE_ITEM.billtoContactName").ToString();
                InvoiceNo.Text =jObj.SelectToken("getInvoice.INVOICE_ITEM.invoiceNo").ToString();
                ShipTo.Text =   jObj.SelectToken("getInvoice.INVOICE_ITEM.shiptoName").ToString();
                ShipAddr.Text = jObj.SelectToken("getInvoice.INVOICE_ITEM.shiptoAddress").ToString();

                var ShipCity =  jObj.SelectToken("getInvoice.INVOICE_ITEM.shiptoCity").ToString();
                var ShipState = jObj.SelectToken("getInvoice.INVOICE_ITEM.shiptoState").ToString();
                var ShipZip =   jObj.SelectToken("getInvoice.INVOICE_ITEM.shiptoZip").ToString();
                Ship_CityStateZip.Text = ShipCity + ", " + ShipState + " " + ShipZip;

                SalesRep.Text = jObj.SelectToken("getInvoice.INVOICE_ITEM.salesRepName").ToString();
                RepPhone.Text = jObj.SelectToken("getInvoice.INVOICE_ITEM.salesRepPhone").ToString();
                Ordered.Text =  jObj.SelectToken("getInvoice.INVOICE_ITEM.orderDate").ToString();
                Invoiced.Text = jObj.SelectToken("getInvoice.INVOICE_ITEM.invoiceDate").ToString();
                Terms.Text =    jObj.SelectToken("getInvoice.INVOICE_ITEM.terms").ToString();
                Note.Text =     jObj.SelectToken("getInvoice.INVOICE_ITEM.comment").ToString();
                Gross.Text =    jObj.SelectToken("getInvoice.INVOICE_ITEM.invoiceGross").ToString();
                Discount.Text = jObj.SelectToken("getInvoice.INVOICE_ITEM.invoiceDiscount").ToString();
                Net.Text =      jObj.SelectToken("getInvoice.INVOICE_ITEM.invoiceNet").ToString();

                // Get Each Line Item
                IList<JToken> Lineresults = jObj["getInvoice"]["INVOICE_ITEM"]["line"].Children().ToList(); //Here don't see ToList()
                int Linecount = Lineresults.Count;

                for (int i = 0; i < Linecount; i++)
                {
                    dataGridView1.Rows.Add();

                    dataGridView1.Rows[i].Cells[0].Value = i+1;
                     dataGridView1.Rows[i].Cells[1].Value = (Lineresults[i].SelectToken("lineQty"));
                     dataGridView1.Rows[i].Cells[2].Value = (Lineresults[i].SelectToken("lineProductNo"));
                     dataGridView1.Rows[i].Cells[3].Value = (Lineresults[i].SelectToken("lineDescription"));
                     dataGridView1.Rows[i].Cells[4].Value = (Lineresults[i].SelectToken("lineUnitPrice"));
                     dataGridView1.Rows[i].Cells[5].Value = (Lineresults[i].SelectToken("lineGross"));
                     dataGridView1.Rows[i].Cells[6].Value = (Lineresults[i].SelectToken("lineDiscount"));
                     dataGridView1.Rows[i].Cells[7].Value = (Lineresults[i].SelectToken("lineExtNet"));

                }
        }
        }

        private void Invoice_Load(object sender, EventArgs e)
        {

        }


        private void NewSearch_Click(object sender, System.EventArgs e)
        {
            CustSearchFormProvider.custsearch.Show();

            List<Form> openForms = new List<Form>();

            foreach (Form f in Application.OpenForms)
                openForms.Add(f);

            foreach (Form f in openForms)
            {
                if (f.Name != "CustomerSearch")
                    f.Close();
            }
        }

        private void Invoice_FormClosing(Object sender, FormClosingEventArgs e)
        {

            this.Dispose();
        }
    }
}
